src/Features/JsonPatch.SystemTextJson/src/JsonPatchDocumentOfT.cs

```diff
@@ -713,19 +713,14 @@ private List<string> GetPathSegments(Expression expr)
 
     private string GetPropertyNameFromMemberExpression(MemberExpression memberExpression)
     {
-        var jsonObjectContract = SerializerOptions.GetTypeInfo(memberExpression.Expression.Type);
-        if (jsonObjectContract != null)
+        var jsonTypeInfo = SerializerOptions.GetTypeInfo(memberExpression.Expression.Type);
+        if (jsonTypeInfo != null)
         {
-            var jsonName = memberExpression.Member.Name;
-            var propNameAttribute = memberExpression.Member.GetCustomAttribute<JsonPropertyNameAttribute>();
-            if (propNameAttribute is not null)
-            {
-                jsonName = propNameAttribute.Name;
-            }
-
-            return jsonObjectContract.Properties
-                .First(jsonProperty => jsonProperty.Name == jsonName)
-                .Name;
+            var memberInfo = memberExpression.Member;
+            var matchingProp = jsonTypeInfo
+                .Properties
+                .First(jsonProp => jsonProp.AttributeProvider is MemberInfo mi && mi == memberInfo);
+            return matchingProp.Name;
```

Comment:
I think this will honor naming policy but it's worth adding a test for it (adding a JsonSerializerOptions.PropertyNamingPolicy to the options and making sure the serialized string path uses that name). Also on the deserialization side, when going from property name -> property info you should make sure naming policy is handled correctly.